File: src/sat_solver.rs
use Cnf;
use cnf_manager::*;
use std::iter::*;
use std::process::exit;

pub struct Luby {
	pub seq : Vec<i32>,
	pub index : i32,
	pub k : i32
}

impl Luby {
	pub fn new() -> Luby {
		Luby {seq : Vec::new(), index : 0, k : 1}
	}

	pub fn next(& mut self) -> i32 {
		self.index += 1;
		if self.index == (1 << self.k) - 1 {
			self.seq.push(1 << (self.k - 1));
			self.k += 1;
		} else {
			let val = self.seq[(self.index - (1 << (self.k - 1))) as usize];
			self.seq.push(val);
		}
		self.seq.last().unwrap().clone()
	}
}

pub struct SatSolver {
	pub cnf_manager : CnfManager,
	pub luby : Luby,
	pub luby_unit : i32,
	pub next_decay : i32,
	pub next_restart : i32
}

impl<'w> SatSolver {
	pub fn new(cnf: &Cnf) -> SatSolver {
		let mut ret = SatSolver {
			cnf_manager : CnfManager::new(cnf),
			luby : Luby::new(),
			luby_unit : 512,
			next_decay : 128,
			next_restart : 0
		};

		ret.next_restart = ret.luby.next()*ret.luby_unit;

		if ret.cnf_manager.decision_level == 0 {
			return ret;
		}


		for i in 1..(ret.cnf_manager.var_count+1) as usize {
			if ret.cnf_manager.vars[i].value == VA::Free {
				if ret.cnf_manager.vars[i].activity[VA::Pos as usize] == 0 && ret.cnf_manager.vars[i].activity[VA::Neg as usize] > 0 {
					ret.cnf_manager.assertLiteral(-(i as i32), ArrTuple::new(), 0);
				} else if ret.cnf_manager.vars[i].activity[VA::Neg as usize] == 0 && ret.cnf_manager.vars[i].activity[VA::Pos as usize] > 0 {
					ret.cnf_manager.assertLiteral((i as i32), ArrTuple::new(), 0);
				}
			}
		}

		for i in 1..(ret.cnf_manager.var_count+1) as usize {
			if ret.cnf_manager.vars[i].value == VA::Free && SCORE(&(i as i32), &ret.cnf_manager) > 0 {
				ret.cnf_manager.var_order.push(i as i32);
				ret.cnf_manager.vars[i].phase = 
					if ret.cnf_manager.vars[i].activity[VA::Pos as usize] > ret.cnf_manager.vars[i].activity[VA::Neg as usize] {
						true
					} else {
						false
					};
			}
		}

		ret.cnf_manager.sort_vars();

		for i in 0..ret.cnf_manager.var_order.len() {
			ret.cnf_manager.var_position[ ret.cnf_manager.var_order[i as usize] as usize ] = i as i32;
		}

		ret.cnf_manager.next_var = 0;
		ret.cnf_manager.next_clause = ret.cnf_manager.clauses.len() as i32 - 1;

		ret
	}

	pub fn selectLiteral(& mut self) -> i32 {
		//println!("selectLiteral");
		let mut x = 0 as i32;
		let last_clause = if self.cnf_manager.next_clause > 256 {
			self.cnf_manager.next_clause - 256
		} else {
			0
		};

		for i in (last_clause..(self.cnf_manager.next_clause + 1)).rev() {
			self.cnf_manager.next_clause = i;

			let mut sat = false;
			let mut ind = self.cnf_manager.clauses[i as usize] as usize;
			while self.cnf_manager.lit_pool[ind] != 0 {
				let lit = self.cnf_manager.lit_pool[ind];
				if lit == 0 {
					break;
				}
				if SET(&lit, &self.cnf_manager) {
					sat = true;
					break;
				}
				ind += 1;
			}
			if sat {
				continue;
			}

			let mut score = -1;
			ind = self.cnf_manager.clauses[i as usize] as usize;
			while self.cnf_manager.lit_pool[ind] != 0 {
				let lit = self.cnf_manager.lit_pool[ind];
				if lit == 0 {
					break;
				}
				if FREE(&lit, &self.cnf_manager) && SCORE(&(VAR(&lit) as i32), &self.cnf_manager) > score {
					x = VAR(&lit) as i32;
					score = SCORE(&x, &self.cnf_manager);
				}
				ind += 1;
			}

			let d = self.cnf_manager.vars[x as usize].activity[VA::Pos as usize]
					- self.cnf_manager.vars[x as usize].activity[VA::Neg as usize];
			if d > 32 {
				return x;
			} else if -d > 32 {
				return -x;
			} else if self.cnf_manager.vars[x as usize].phase {
				return x;
			} else {
				return -x;
			}
		}

		for i in (self.cnf_manager.next_var as usize)..self.cnf_manager.var_order.len() {
			if self.cnf_manager.vars[self.cnf_manager.var_order[i] as usize].value == VA::Free {
				x = self.cnf_manager.var_order[i];
				self.cnf_manager.next_var = (i + 1) as i32;

				let d = self.cnf_manager.vars[x as usize].activity[VA::Pos as usize]
					- self.cnf_manager.vars[x as usize].activity[VA::Neg as usize];
				if d > 32 {
					return x;
				} else if -d > 32 {
					return -x;
				} else if self.cnf_manager.vars[x as usize].phase {
					return x;
				} else {
					return -x;
				}
			}
		}
		0
	}

	pub fn run(&mut self) -> bool {
		//println!("run {}", self.cnf_manager.decision_level);
		if self.cnf_manager.decision_level == 0 {
			return false;
		}
		let mut lit = self.selectLiteral();
		while lit != 0 {
			//println!("selected literal: {}", lit);
			if !self.cnf_manager.decide(lit) {
				loop {
					//println!("aLevel {}", self.cnf_manager.assertion_level);
					if self.cnf_manager.assertion_level == 0 {
						return false;
					}

					if self.cnf_manager.conflict_count == self.next_decay {
						self.next_decay += 128;
						self.cnf_manager.scoreDecay();
					}

					self.cnf_manager.next_clause = self.cnf_manager.clauses.len() as i32 - 1;

					if self.cnf_manager.conflict_count == self.next_restart {
						self.cnf_manager.restart_count += 1;
						self.next_restart += self.luby.next() * self.luby_unit;
						self.cnf_manager.backtrack(1);
						if self.cnf_manager.decision_level != self.cnf_manager.assertion_level {
							break;
						}
					} else {
						let level = self.cnf_manager.assertion_level;
						self.cnf_manager.backtrack(level);
					}

					if self.cnf_manager.assertCL() {
						break;
					}
				}
			}
			lit = self.selectLiteral();
		}
		if !self.verifySolution() {
			println!("ERROR");
			exit(0);
		}
		true
	}

	pub fn verifySolution(&self) -> bool {
		let ref pool = self.cnf_manager.lit_pool;
		let mut i = 0;
		while i < self.cnf_manager.lit_pool_size_orig {
			let mut satisfied = false;
			while pool[i] != 0 {
				let ref lit = pool[i];
				i += 1;
				if SET(&lit, &self.cnf_manager) {
					satisfied = true;
					while pool[i] != 0 {
						i += 1;
					}
					break;
				}
			}
			i += 1;
			if !satisfied {
				return false;
			}
		}
		true
	}

	pub fn printSolution(&self) -> () {
		for i in 1..(self.cnf_manager.var_count+1) as usize {
			let ref vars = self.cnf_manager.vars;
			if vars[i].value == VA::Pos {
				print!("{} ", i);
			} else {
				print!("-{} ", i);
			}
		}
		println!("0");
	}

	pub fn printStats(&self) -> () {
		let ref m = self.cnf_manager;
		println!("c {} decisions, {} conflicts, {} restarts", m.decision_count, m.conflict_count, m.restart_count);
	}
}

File: cpp/SatSolver.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <algorithm>
#include <functional>
#include "SatSolver.h"

#define HALFLIFE 128
#define _DT 32        // RSAT phase selection threshold

SatSolver::SatSolver(Cnf &cnf) : CnfManager(cnf) {
    // initialize parameters
    nextRestart = luby.next() * (lubyUnit = 512);
    nextDecay = HALFLIFE;

    // assertUnitClauses has failed
    if (dLevel == 0) return;

    // assert pure literals
    for (int i = 1; i <= (int) vc; i++)
        if (vars[i].value == _FREE) {
            if (vars[i].activity[_POSI] == 0 && vars[i].activity[_NEGA] > 0)
                // ante is NULL, as opposed to empty clause for implied literals
                assertLiteral(-i, NULL);
            else if (vars[i].activity[_NEGA] == 0 && vars[i].activity[_POSI] > 0)
                assertLiteral(i, NULL);
        }

    // initialize varOrder
    nVars = 0;
    for (unsigned i = 1; i <= vc; i++)
        if (vars[i].value == _FREE && SCORE(i) > 0) {
            varOrder[nVars++] = i;
            vars[i].phase = (vars[i].activity[_POSI] > vars[i].activity[_NEGA]) ? _POSI : _NEGA;
        }
    sort(varOrder, varOrder + nVars, [&](unsigned a, unsigned b) -> bool { return SCORE(a) > SCORE(b); });
    for (unsigned i = 0; i < nVars; i++) varPosition[varOrder[i]] = i;
    nextVar = 0;
    printf("next clause: %d\n", clauses.size() - 1);
    nextClause = clauses.size() - 1;
}

int SatSolver::selectLiteral() {
    unsigned x = 0;

    // pick best var in unsatisfied conflict clause nearest to top of stack
    // but only search 256 clauses
    int lastClause = nextClause > 256 ? (nextClause - 256) : 0;
    for (int i = nextClause; i >= lastClause; i--) {
        int *p = clauses[nextClause = i];

        // skip satisfied clauses
        bool sat = false;
        for (; (*p); p++)
            if (SET(*p)) {
                sat = true;
                break;
            }
        if (sat) continue;

        // traverse again, find best variable of clause
        int score = -1;
        for (p = clauses[i]; (*p); p++)
            if (FREE(*p) && ((int) SCORE(VAR(*p))) > score) {
                x = VAR(*p);
                score = SCORE(x);
            }

        // RSAT phase selection
        int d = vars[x].activity[_POSI] - vars[x].activity[_NEGA];
        if (d > _DT) return x; else if (-d > _DT) return -x;
        else return (vars[x].phase == _POSI) ? (x) : -(int) (x);
    }

    // fall back to VSIDS
    for (unsigned i = nextVar; i < nVars; i++) {
        if (vars[varOrder[i]].value == _FREE) {
            x = varOrder[i];
            nextVar = i + 1;

            // RSAT phase selection
            int d = vars[x].activity[_POSI] - vars[x].activity[_NEGA];
            if (d > _DT) return x; else if (-d > _DT) return -x;
            else return (vars[x].phase == _POSI) ? (x) : -(int) (x);
        }
    }
    return 0;
}

bool SatSolver::run() {
    if (dLevel == 0) return false;        // assertUnitClauses has failed
    for (int lit; (lit = selectLiteral());) { // pick decision literal
        if (!decide(lit))
            do {        // decision/conflict
                // conflict has occurred in dLevel 1, unsat
                if (aLevel == 0) return false;

                // score decay
                if (nConflicts == nextDecay) {
                    nextDecay += HALFLIFE;
                    scoreDecay();
                }

                // rewind to top of clause stack
                nextClause = clauses.size() - 1;

                // restart at dLevel 1
                if (nConflicts == nextRestart) {
                    nRestarts++;
                    nextRestart += luby.next() * lubyUnit;
                    backtrack(1);
                    if (dLevel != aLevel) break;

                    // partial restart at aLevel
                } else backtrack(aLevel);
            } while (!assertCL());        // assert conflict literal
    }
    if (!verifySolution()) {
        printf("s UNKOWN\n");
        exit(0);
    }
    return true;
}

bool SatSolver::verifySolution() {
    int lit, *pool = litPools[0];
    for (unsigned i = 0; i < litPoolSizeOrig;) {
        bool satisfied = false;
        while ((lit = pool[i++]))
            if (SET(lit)) {
                satisfied = true;
                while (pool[i++]);
                break;
            }
        if (!satisfied) return false;
    }
    return true;
}

void SatSolver::printSolution(FILE *ofp) {
    for (unsigned i = 1; i <= vc; i++)
        if (vars[i].value == _POSI) fprintf(ofp, "%d ", i);
        else if (vars[i].value == _NEGA) fprintf(ofp, "-%d ", i);
    fprintf(ofp, "0\n");
}

void SatSolver::printStats() {
    printf("c %d decisions, %d conflicts, %d restarts\n", nDecisions, nConflicts, nRestarts);
}
